prometheus_hardware_exporter/collector.py

```diff
@@ -874,21 +875,141 @@ def specifications(self) -> List[Specification]:
         """Load redfish data."""
-        payloads = []
-        service_status = self.redfish_helper.discover()
-        payloads.append(Payload(name="redfish_service_available", value=float(service_status)))
-
-        sensor_data = self.redfish_helper.get_sensor_data()
-        if not sensor_data:
-            logger.error("Failed to get sensor data via redfish.")
-            payloads.append(Payload(name="redfish_call_success", value=0.0))
-            return payloads
+        payloads: List[Payload] = []
+        with RedfishHelper(self.config) as redfish_helper:
+            service_status = self.discover_redfish_services()
+            payloads.append(Payload(name="redfish_service_available", value=float(service_status)))
+
+            processor_count: Dict[str, int]
+            processor_data: Dict[str, List]
+            storage_controller_count: Dict[str, int]
+            storage_controller_data: Dict[str, List]
+            storage_drive_count: Dict[str, int]
+            storage_drive_data: Dict[str, List]
+            memory_dimm_count: Dict[str, int]
+            memory_dimm_data: Dict[str, List]
+
+            sensor_data: Dict[str, List] = redfish_helper.get_sensor_data()
+            processor_count, processor_data = redfish_helper.get_processor_data()
+            (
+                storage_controller_count,
+                storage_controller_data,
+            ) = redfish_helper.get_storage_controller_data()
+            network_adapter_count: Dict[str, Any] = redfish_helper.get_network_adapter_data()
+            chassis_data: Dict[str, Dict] = redfish_helper.get_chassis_data()
+            storage_drive_count, storage_drive_data = redfish_helper.get_storage_drive_data()
+            memory_dimm_count, memory_dimm_data = redfish_helper.get_memory_dimm_data()
+            smart_storage_health_data: Dict[
+                str, Any
+            ] = redfish_helper.get_smart_storage_health_data()
+
+        metrics = {
+            "sensor_data": sensor_data,
+            "processor_data": processor_data,
+            "processor_count": processor_count,
+            "storage_controller_data": storage_controller_data,
+            "storage_controller_count": storage_controller_count,
+            "network_adapter_count": network_adapter_count,
+            "chassis_data": chassis_data,
+            "storage_drive_data": storage_drive_data,
+            "storage_drive_count": storage_drive_count,
+            "memory_dimm_data": memory_dimm_data,
+            "memory_dimm_count": memory_dimm_count,
+            "smart_storage_health_data": smart_storage_health_data,
+        }
+
+        for metric_name, metric_data in metrics.items():
+            if not metric_data:
+                logger.error("Failed to get %s via redfish", metric_name)
 
         payloads.append(Payload(name="redfish_call_success", value=1.0))
+
+        payloads = self._create_sensor_metric_payload(metrics["sensor_data"], payloads)
```

Comment:
It's better to write it something like this
```suggestion
        payloads.extend (self._create_sensor_metric_payload(metrics["sensor_data"]))
```
It will decoupling of each internal function.

Reply:
This is a nice idea :+1: